File: src/main/java/com/multithreading/ThreadSynchronization.java
package com.multithreading;

/**
 * Thread Synchronization Example
 * Demonstrates thread synchronization using synchronized keyword to prevent race conditions.
 */
public class ThreadSynchronization {

    /**
     * Thread-safe counter using synchronized methods.
     */
    static class SafeCounter {
        private int value = 0;

        public synchronized void increment() {
            int current = value;
            try {
                Thread.sleep(1); // Simulate some processing
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            value = current + 1;
        }

        public synchronized int getValue() {
            return value;
        }
    }

    /**
     * Unsafe counter without synchronization (demonstrates race conditions).
     */
    static class UnsafeCounter {
        private int value = 0;

        public void increment() {
            int current = value;
            try {
                Thread.sleep(1); // Simulate some processing
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            value = current + 1;
        }

        public int getValue() {
            return value;
        }
    }

    /**
     * Worker that increments a counter.
     */
    static class IncrementWorker extends Thread {
        private final SafeCounter safeCounter;
        private final UnsafeCounter unsafeCounter;
        private final int iterations;

        public IncrementWorker(SafeCounter safeCounter, UnsafeCounter unsafeCounter, int iterations) {
            this.safeCounter = safeCounter;
            this.unsafeCounter = unsafeCounter;
            this.iterations = iterations;
        }

        @Override
        public void run() {
            for (int i = 0; i < iterations; i++) {
                if (safeCounter != null) {
                    safeCounter.increment();
                }
                if (unsafeCounter != null) {
                    unsafeCounter.increment();
                }
            }
        }
    }

    public static void runExample() throws InterruptedException {
        System.out.println("=== Thread Synchronization Example ===\n");

        int numThreads = 10;
        int iterations = 100;

        // Test with unsafe counter
        System.out.println("1. Unsafe Counter (Race Condition):");
        UnsafeCounter unsafeCounter = new UnsafeCounter();
        Thread[] unsafeThreads = new Thread[numThreads];

        for (int i = 0; i < numThreads; i++) {
            unsafeThreads[i] = new IncrementWorker(null, unsafeCounter, iterations);
            unsafeThreads[i].start();
        }

        for (Thread thread : unsafeThreads) {
            thread.join();
        }

        System.out.println("Expected: " + (numThreads * iterations) + ", Got: " + unsafeCounter.getValue());

        // Test with safe counter
        System.out.println("\n2. Safe Counter (With Synchronization):");
        SafeCounter safeCounter = new SafeCounter();
        Thread[] safeThreads = new Thread[numThreads];

        for (int i = 0; i < numThreads; i++) {
            safeThreads[i] = new IncrementWorker(safeCounter, null, iterations);
            safeThreads[i].start();
        }

        for (Thread thread : safeThreads) {
            thread.join();
        }

        System.out.println("Expected: " + (numThreads * iterations) + ", Got: " + safeCounter.getValue());
        System.out.println();
    }

    public static void main(String[] args) throws InterruptedException {
        runExample();
    }
}

File: thread_synchronization.py
"""
Thread Synchronization Example
Demonstrates thread synchronization using locks to prevent race conditions.
"""
import threading
import time


class Counter:
    """A thread-safe counter using locks."""
    
    def __init__(self):
        self.value = 0
        self.lock = threading.Lock()
    
    def increment(self):
        """Increment counter in a thread-safe manner."""
        with self.lock:
            current = self.value
            time.sleep(0.001)  # Simulate some processing
            self.value = current + 1
    
    def get_value(self):
        """Get current counter value."""
        with self.lock:
            return self.value


class UnsafeCounter:
    """A counter without thread synchronization (demonstrates race conditions)."""
    
    def __init__(self):
        self.value = 0
    
    def increment(self):
        """Increment counter without thread safety."""
        current = self.value
        time.sleep(0.001)  # Simulate some processing
        self.value = current + 1
    
    def get_value(self):
        """Get current counter value."""
        return self.value


def increment_worker(counter, iterations):
    """Worker that increments a counter multiple times."""
    for _ in range(iterations):
        counter.increment()


def run_synchronization_example():
    """Run thread synchronization example comparing safe and unsafe counters."""
    print("=== Thread Synchronization Example ===")
    
    # Test with unsafe counter
    print("\n1. Unsafe Counter (Race Condition):")
    unsafe_counter = UnsafeCounter()
    threads = []
    iterations = 100
    
    for i in range(10):
        thread = threading.Thread(target=increment_worker, args=(unsafe_counter, iterations))
        threads.append(thread)
        thread.start()
    
    for thread in threads:
        thread.join()
    
    print(f"Expected: {10 * iterations}, Got: {unsafe_counter.get_value()}")
    
    # Test with safe counter
    print("\n2. Safe Counter (With Lock):")
    safe_counter = Counter()
    threads = []
    
    for i in range(10):
        thread = threading.Thread(target=increment_worker, args=(safe_counter, iterations))
        threads.append(thread)
        thread.start()
    
    for thread in threads:
        thread.join()
    
    print(f"Expected: {10 * iterations}, Got: {safe_counter.get_value()}")
    print()


if __name__ == "__main__":
    run_synchronization_example()
